File: lambda_function/handler.py
"""
AWS Lambda handler: Databricks PRD → Supabase incremental sync.

Triggered by EventBridge schedule. Uses OAuth M2M (Service Principal)
for Databricks and REST API for Supabase.

New records get estado='pendiente'; existing records get data updated
without touching estado_transportista.
"""
import json
import logging
import os
from datetime import datetime, timezone
from typing import Any

import httpx
from databricks import sql
from databricks.sdk.core import Config, oauth_service_principal
from id_generator import generate_transportista_id

logger = logging.getLogger()
logger.setLevel(logging.INFO)

# ---------------------------------------------------------------------------
# CONFIG (from Lambda environment variables)
# ---------------------------------------------------------------------------

DATABRICKS_HOST = os.environ.get("DATABRICKS_HOST", "")
DATABRICKS_HTTP_PATH = os.environ.get("DATABRICKS_HTTP_PATH", "")
DATABRICKS_CLIENT_ID = os.environ.get("DATABRICKS_CLIENT_ID", "")
DATABRICKS_CLIENT_SECRET = os.environ.get("DATABRICKS_CLIENT_SECRET", "")

SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

DBX_TABLE = os.environ.get("DBX_TABLE", "prod.gldprd.dim_app_transportistas")
SUPABASE_TABLE = os.environ.get("SUPABASE_TABLE", "transportistas")
WATERMARK_TABLE = "etl_watermarks"
BATCH_SIZE = int(os.environ.get("BATCH_SIZE", "500"))

HEADERS = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type": "application/json",
}

# ---------------------------------------------------------------------------
# DATABRICKS (OAuth M2M)
# ---------------------------------------------------------------------------

def _get_connection():
    host = DATABRICKS_HOST
    cid = DATABRICKS_CLIENT_ID
    csecret = DATABRICKS_CLIENT_SECRET

    def _credential_provider():
        config = Config(
            host=f"https://{host}",
            client_id=cid,
            client_secret=csecret,
        )
        return oauth_service_principal(config)

    return sql.connect(
        server_hostname=host,
        http_path=DATABRICKS_HTTP_PATH,
        credentials_provider=_credential_provider,
    )


def fetch_databricks() -> list[tuple]:
    query = f"""
    SELECT
      codigo_transportista, ruc, nombre_transportista,
      telefono, email, estado_transportista,
      placas, tipos_vehiculo
    FROM {DBX_TABLE}
    """
    logger.info("Querying Databricks PRD: %s", DBX_TABLE)
    with _get_connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(query)
            rows = cursor.fetchall()
            row_tuples = [tuple(r) if not isinstance(r, tuple) else r for r in rows]
    logger.info("Fetched %d row(s) from Databricks.", len(row_tuples))
    return row_tuples

# ---------------------------------------------------------------------------
# SUPABASE REST
# ---------------------------------------------------------------------------

def supabase_get(table: str, params: dict | None = None) -> httpx.Response:
    return httpx.get(
        f"{SUPABASE_URL}/rest/v1/{table}",
        headers=HEADERS,
        params=params or {},
        timeout=30,
    )


def supabase_post(table: str, rows: list[dict]) -> httpx.Response:
    return httpx.post(
        f"{SUPABASE_URL}/rest/v1/{table}",
        headers={**HEADERS, "Prefer": "return=minimal"},
        json=rows,
        timeout=60,
    )


def supabase_upsert(table: str, rows: list[dict]) -> httpx.Response:
    return httpx.post(
        f"{SUPABASE_URL}/rest/v1/{table}",
        headers={**HEADERS, "Prefer": "resolution=merge-duplicates"},
        json=rows,
        timeout=60,
    )

# ---------------------------------------------------------------------------
# HELPERS
# ---------------------------------------------------------------------------

def get_existing_codigos() -> set[str]:
    codigos: set[str] = set()
    offset = 0
    batch = 1000
    while True:
        r = httpx.get(
            f"{SUPABASE_URL}/rest/v1/{SUPABASE_TABLE}",
            headers={**HEADERS, "Range": f"{offset}-{offset + batch - 1}"},
            params={"select": "codigo_transportista"},
            timeout=30,
        )
        if r.status_code not in (200, 206):
            raise RuntimeError(f"Failed to read existing codes: {r.status_code}")
        data = r.json()
        if not data:
            break
        for row in data:
            codigos.add(row["codigo_transportista"])
        if len(data) < batch:
            break
        offset += batch
    return codigos


def _serialize_array(v: Any):
    if v is None:
        return None
    if hasattr(v, "tolist"):
        v = v.tolist()
    if isinstance(v, str):
        v = v.strip()
        if v.startswith("["):
            try:
                v = json.loads(v)
            except (json.JSONDecodeError, ValueError):
                return [v]
        elif v.startswith("{") and v.endswith("}"):
            v = [item.strip().strip('"') for item in v[1:-1].split(",") if item.strip()]
        else:
            return [v] if v else None
    if isinstance(v, (list, tuple)):
        return [str(item).strip() if item is not None else None for item in v]
    return [v] if v else None


def deduplicate_rows(rows: list[tuple]) -> list[tuple]:
    seen: dict[str, tuple] = {}
    for row in rows:
        codigo = str(row[0]).strip() if row[0] else None
        if codigo:
            seen[codigo] = row
    dupes = len(rows) - len(seen)
    if dupes:
        logger.warning("Removed %d duplicate codigo_transportista from Databricks data.", dupes)
    return list(seen.values())


def _clean_unique_field(value: Any) -> str | None:
    """Return None for empty/whitespace-only strings (safe for UNIQUE constraints)."""
    if value is None:
        return None
    s = str(value).strip()
    return s if s else None


def split_new_and_existing(
    rows: list[tuple], existing_codigos: set[str], now: str
) -> list[dict]:
    new_rows: list[dict] = []
    seen_emails: set[str] = set()
    seen_phones: set[str] = set()

    for row in rows:
        codigo = str(row[0]).strip() if row[0] else None
        ruc = str(row[1]).strip() if row[1] else None
        if not codigo or not ruc:
            continue

        email = _clean_unique_field(row[4])
        telefono = _clean_unique_field(row[3])

        if email and email in seen_emails:
            email = None
        if telefono and telefono in seen_phones:
            telefono = None

        if email:
            seen_emails.add(email)
        if telefono:
            seen_phones.add(telefono)

        record = {
            "codigo_transportista": codigo,
            "ruc": ruc,
            "nombre_transportista": row[2],
            "telefono": telefono,
            "email": email,
            "placas": _serialize_array(row[6]),
            "tipos_vehiculo": _serialize_array(row[7]),
            "_ingested_at": now,
        }

        if codigo not in existing_codigos:
            record["transportista_id"] = generate_transportista_id(codigo)
            record["estado_transportista"] = "pendiente"
            new_rows.append(record)

    return new_rows

# ---------------------------------------------------------------------------
# SYNC LOGIC
# ---------------------------------------------------------------------------

def run_sync() -> dict:
    sync_start = datetime.now(timezone.utc)
    now = sync_start.isoformat()

    # Validate Supabase tables
    r = supabase_get(SUPABASE_TABLE, {"select": "codigo_transportista", "limit": "1"})
    if r.status_code != 200:
        raise RuntimeError(f"Cannot access '{SUPABASE_TABLE}': {r.status_code}")
    logger.info("Table '%s' accessible.", SUPABASE_TABLE)

    r_wm = supabase_get(WATERMARK_TABLE, {"select": "table_name", "limit": "1"})
    if r_wm.status_code != 200:
        raise RuntimeError(f"Cannot access '{WATERMARK_TABLE}': {r_wm.status_code}")

    # Get existing codes
    existing = get_existing_codigos()
    count_before = len(existing)
    logger.info("Found %d existing records in Supabase.", count_before)

    # Fetch from Databricks
    dbx_rows = fetch_databricks()
    if not dbx_rows:
        logger.info("No rows from Databricks. Nothing to sync.")
        return {"status": "no_data", "inserted": 0, "total": count_before}

    # Deduplicate
    dbx_rows = deduplicate_rows(dbx_rows)

    # Filter new only
    new_rows = split_new_and_existing(dbx_rows, existing, now)
    logger.info("New: %d | Already in Supabase: %d (skipped)", len(new_rows), len(dbx_rows) - len(new_rows))

    # Insert new
    inserted = 0
    if new_rows:
        for i in range(0, len(new_rows), BATCH_SIZE):
            batch = new_rows[i : i + BATCH_SIZE]
            r = supabase_post(SUPABASE_TABLE, batch)
            if r.status_code not in (200, 201):
                raise RuntimeError(f"Insert failed: {r.status_code} {r.text[:300]}")
            inserted += len(batch)
            logger.info("[NEW] Batch %d: %d rows", i // BATCH_SIZE + 1, len(batch))
    else:
        logger.info("No new records to insert.")

    # Post-sync verification
    post_codigos = get_existing_codigos()
    count_after = len(post_codigos)
    expected = count_before + inserted
    if count_after != expected:
        logger.warning("Post-sync mismatch: expected %d, found %d.", expected, count_after)
    else:
        logger.info("Post-sync OK: %d rows (was %d, +%d new).", count_after, count_before, inserted)

    # Watermark
    sync_time = datetime.now(timezone.utc).isoformat()
    r = supabase_upsert(WATERMARK_TABLE, [{
        "table_name": SUPABASE_TABLE,
        "last_timestamp": sync_time,
    }])
    if r.status_code in (200, 201):
        logger.info("Watermark updated: %s", sync_time)
    else:
        logger.warning("Watermark update failed: %s", r.status_code)

    return {
        "status": "success",
        "inserted": inserted,
        "total": count_after,
        "sync_timestamp": sync_time,
    }

# ---------------------------------------------------------------------------
# LAMBDA HANDLER
# ---------------------------------------------------------------------------

def lambda_handler(event, context):
    logger.info("Event: %s", json.dumps(event, default=str))

    try:
        result = run_sync()
        logger.info("Sync complete: %s", json.dumps(result))
        return {
            "statusCode": 200,
            "body": json.dumps(result),
        }
    except Exception as e:
        logger.error("Sync failed: %s", e, exc_info=True)
        return {
            "statusCode": 500,
            "body": json.dumps({"status": "error", "message": str(e)}),
        }

File: lambda_function/handler.js
const fs = require("fs");
const path = require("path");
const { S3Client, PutObjectCommand } = require("@aws-sdk/client-s3");
const { v5: uuidv5 } = require("uuid");

const DATABRICKS_HOST = process.env.DATABRICKS_HOST || "";
const DATABRICKS_HTTP_PATH = process.env.DATABRICKS_HTTP_PATH || "";
const DATABRICKS_CLIENT_ID = process.env.DATABRICKS_CLIENT_ID || "";
const DATABRICKS_CLIENT_SECRET = process.env.DATABRICKS_CLIENT_SECRET || "";
const DATABRICKS_QAS_HOST = process.env.DATABRICKS_QAS_HOST || "";
const DATABRICKS_QAS_HTTP_PATH = process.env.DATABRICKS_QAS_HTTP_PATH || "";
const DATABRICKS_QAS_TOKEN = process.env.DATABRICKS_QAS_TOKEN || "";

const SUPABASE_URL = (process.env.SUPABASE_URL || "").replace(/\/$/, "");
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || "";
const SUPABASE_SECONDARY_URL = (process.env.SUPABASE_SECONDARY_URL || "").replace(/\/$/, "");
const SUPABASE_SECONDARY_KEY = process.env.SUPABASE_SECONDARY_SERVICE_ROLE_KEY || "";

const WATERMARK_TABLE = "etl_watermarks";
const BATCH_SIZE = parseInt(process.env.BATCH_SIZE || "500", 10);
const ETL_LOGS_BUCKET = process.env.ETL_LOGS_BUCKET || "";
const UUID_NAMESPACE = "b8f9e3a1-7c2d-4f5e-9a1b-3c4d5e6f7a8b";
const PIPELINES_CONFIG_PATH =
  process.env.PIPELINES_CONFIG_PATH || path.join(__dirname, "pipelines.json");

const s3Client = new S3Client({});

function logInfo(message, extra) {
  if (extra !== undefined) {
    console.log(`${message}: ${JSON.stringify(extra)}`);
  } else {
    console.log(message);
  }
}

function resolveSupabaseForPipeline(cfg) {
  const profile = cfg.supabase_profile || "default";
  if (profile === "secondary") {
    if (!SUPABASE_SECONDARY_URL || !SUPABASE_SECONDARY_KEY) {
      throw new Error(
        "Pipeline uses supabase_profile=secondary: set SUPABASE_SECONDARY_URL and SUPABASE_SECONDARY_SERVICE_ROLE_KEY on the Lambda"
      );
    }
    return { baseUrl: SUPABASE_SECONDARY_URL, serviceKey: SUPABASE_SECONDARY_KEY };
  }
  if (!SUPABASE_URL || !SUPABASE_KEY) {
    throw new Error("Default Supabase (SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY) not configured");
  }
  return { baseUrl: SUPABASE_URL, serviceKey: SUPABASE_KEY };
}

function makeSupabaseClient(baseUrl, serviceKey) {
  const u = (baseUrl || "").replace(/\/$/, "");
  const getHeaders = (prefer) => ({
    apikey: serviceKey,
    Authorization: `Bearer ${serviceKey}`,
    "Content-Type": "application/json",
    ...(prefer ? { Prefer: prefer } : {}),
  });
  return {
    get(table, params = {}, extraHeaders = {}) {
      const url = new URL(`${u}/rest/v1/${table}`);
      Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v));
      return fetch(url, {
        headers: { ...getHeaders(), ...extraHeaders },
      });
    },
    post(table, rows, prefer) {
      return fetch(`${u}/rest/v1/${table}`, {
        method: "POST",
        headers: getHeaders(`resolution=ignore-duplicates,return=minimal${prefer ? `,${prefer}` : ""}`),
        body: JSON.stringify(rows),
      });
    },
    upsert(table, rows, conflictKey) {
      const url = new URL(`${u}/rest/v1/${table}`);
      if (conflictKey) {
        url.searchParams.set("on_conflict", conflictKey);
      }
      return fetch(url, {
        method: "POST",
        headers: getHeaders("resolution=merge-duplicates"),
        body: JSON.stringify(rows),
      });
    },
  };
}

function generateTransportistaId(codigoTransportista) {
  return uuidv5(String(codigoTransportista).trim(), UUID_NAMESPACE);
}

function loadPipelinesConfig() {
  const raw = fs.readFileSync(PIPELINES_CONFIG_PATH, "utf8");
  return JSON.parse(raw);
}

function getPipelineConfig(event) {
  const pipelineName = event?.pipeline_name || process.env.DEFAULT_PIPELINE || "transportistas";
  const all = loadPipelinesConfig();
  const cfg = all[pipelineName];
  if (!cfg) {
    throw new Error(`Unknown pipeline_name '${pipelineName}'`);
  }
  return cfg;
}

function cleanUniqueField(value) {
  if (value === null || value === undefined) return null;
  const s = String(value).trim();
  return s || null;
}

function sourceIndexMap(pipelineConfig) {
  return Object.fromEntries(
    pipelineConfig.column_mapping.map((m, i) => [m.source, i])
  );
}

function conflictSourceColumnIndex(pipelineConfig) {
  const src = pipelineConfig.column_mapping.find((m) => m.target === pipelineConfig.conflict_key);
  if (!src) return 0;
  return pipelineConfig.column_mapping.findIndex((m) => m.source === src.source);
}

function deduplicateRows(rows, pipelineConfig) {
  const idx = conflictSourceColumnIndex(pipelineConfig);
  const seen = new Map();
  for (const row of rows) {
    const k = row[idx] != null ? String(row[idx]).trim() : null;
    if (k) {
      seen.set(k, row);
    }
  }
  const dupes = rows.length - seen.size;
  if (dupes > 0) {
    console.warn(`Removed ${dupes} duplicate rows on conflict key from Databricks.`);
  }
  return [...seen.values()];
}

function splitRowsByMode(rows, pipelineConfig, existingKeys, now) {
  const rowMode = pipelineConfig.row_mode || "transportistas";
  if (rowMode === "generic") {
    return splitRowsGeneric(rows, pipelineConfig, existingKeys, now);
  }
  return splitRowsTransportistas(rows, pipelineConfig, existingKeys, now);
}

function splitRowsGeneric(rows, pipelineConfig, existingKeys, now) {
  const sourceIndex = sourceIndexMap(pipelineConfig);
  const conflictKey = pipelineConfig.conflict_key;
  const srcForConflict = pipelineConfig.column_mapping.find(
    (m) => m.target === conflictKey
  )?.source;
  const defaultRequire = srcForConflict ? [srcForConflict] : [];
  const requireNonNull = pipelineConfig.require_non_null || defaultRequire;
  const includeIngested = pipelineConfig.include_ingested_at !== false;

  const newRows = [];
  const allRows = [];

  for (const row of rows) {
    let skip = false;
    for (const src of requireNonNull) {
      const vi = sourceIndex[src];
      const v = vi !== undefined ? row[vi] : undefined;
      if (v === null || v === undefined || String(v).trim() === "") {
        skip = true;
        break;
      }
    }
    if (skip) continue;

    const record = {};
    for (const mapping of pipelineConfig.column_mapping) {
      const i = sourceIndex[mapping.source];
      record[mapping.target] = i !== undefined ? row[i] : null;
    }
    if (pipelineConfig.defaults) {
      Object.assign(record, pipelineConfig.defaults);
    }
    if (pipelineConfig.null_coalesce) {
      for (const [k, v] of Object.entries(pipelineConfig.null_coalesce)) {
        if (record[k] == null || record[k] === "") {
          record[k] = v;
        }
      }
    }
    const idS = pipelineConfig.id_strategy;
    if (idS?.type === "uuid5_from_source" && idS.source_column && idS.column) {
      const si = sourceIndex[idS.source_column];
      const ns = idS.namespace || UUID_NAMESPACE;
      if (si !== undefined && row[si] != null) {
        record[idS.column] = uuidv5(String(row[si]).trim(), ns);
      }
    }
    if (includeIngested) {
      const tsCol = pipelineConfig.sync_timestamp_column || "_ingested_at";
      record[tsCol] = now;
    }

    const ck = record[conflictKey];
    const ckStr = ck != null ? String(ck) : "";
    allRows.push(record);
    if (ckStr && !existingKeys.has(ckStr)) {
      newRows.push(record);
    }
  }
  return { newRows, allRows };
}

function splitRowsTransportistas(rows, pipelineConfig, existingKeys, now) {
  const newRows = [];
  const allRows = [];
  const seenEmails = new Set();
  const seenPhones = new Set();
  const sourceIndex = sourceIndexMap(pipelineConfig);
  const conflictKey = pipelineConfig.conflict_key;

  for (const row of rows) {
    const codigo = row[sourceIndex.codigo_transportista]
      ? String(row[sourceIndex.codigo_transportista]).trim()
      : null;
    const ruc = row[sourceIndex.ruc] ? String(row[sourceIndex.ruc]).trim() : null;
    if (!codigo || !ruc) continue;

    let email =
      sourceIndex.email !== undefined ? cleanUniqueField(row[sourceIndex.email]) : null;
    let telefono =
      sourceIndex.telefono !== undefined
        ? cleanUniqueField(row[sourceIndex.telefono])
        : null;

    if (email && seenEmails.has(email)) email = null;
    if (telefono && seenPhones.has(telefono)) telefono = null;

    if (email) seenEmails.add(email);
    if (telefono) seenPhones.add(telefono);

    const record = {};
    for (const mapping of pipelineConfig.column_mapping) {
      const idx = sourceIndex[mapping.source];
      let value = idx !== undefined ? row[idx] : null;
      if (mapping.target === "email") value = email;
      if (mapping.target === "telefono") value = telefono;
      record[mapping.target] = value;
    }
    if (pipelineConfig.defaults) {
      Object.assign(record, pipelineConfig.defaults);
    }
    if (pipelineConfig.id_strategy?.type === "uuid5_codigo_transportista") {
      record[pipelineConfig.id_strategy.column] = generateTransportistaId(codigo);
    }
    record._ingested_at = now;

    allRows.push(record);
    const ck = record[conflictKey] != null ? String(record[conflictKey]) : "";
    if (ck && !existingKeys.has(ck)) {
      newRows.push(record);
    }
  }
  return { newRows, allRows };
}

function resolveDatabricksContext(pipelineConfig) {
  const profile = pipelineConfig.databricks_profile || "prd";
  if (profile === "qas") {
    if (!DATABRICKS_QAS_HOST || !DATABRICKS_QAS_HTTP_PATH || !DATABRICKS_QAS_TOKEN) {
      throw new Error(
        "Pipeline uses databricks_profile=qas: set DATABRICKS_QAS_HOST, DATABRICKS_QAS_HTTP_PATH, DATABRICKS_QAS_TOKEN on the Lambda"
      );
    }
    return {
      host: DATABRICKS_QAS_HOST,
      httpPath: DATABRICKS_QAS_HTTP_PATH,
      auth: "pat",
      pat: DATABRICKS_QAS_TOKEN
    };
  }
  if (!DATABRICKS_HOST || !DATABRICKS_HTTP_PATH) {
    throw new Error("Databricks PRD (DATABRICKS_HOST / DATABRICKS_HTTP_PATH) not configured");
  }
  return { host: DATABRICKS_HOST, httpPath: DATABRICKS_HTTP_PATH, auth: "oauth" };
}

async function getDatabricksOAuthToken() {
  const tokenUrl = `https://${DATABRICKS_HOST}/oidc/v1/token`;
  const auth = Buffer.from(`${DATABRICKS_CLIENT_ID}:${DATABRICKS_CLIENT_SECRET}`).toString(
    "base64"
  );
  const body = new URLSearchParams({
    grant_type: "client_credentials",
    scope: "all-apis",
  });

  const res = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      Authorization: `Basic ${auth}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: body.toString(),
  });

  if (!res.ok) {
    const txt = await res.text();
    throw new Error(`Databricks OAuth failed (${res.status}): ${txt.slice(0, 300)}`);
  }

  const data = await res.json();
  return data.access_token;
}

async function fetchDatabricks(pipelineConfig) {
  const sourceColumns = pipelineConfig.column_mapping.map((m) => m.source).join(", ");
  const statement = `SELECT ${sourceColumns} FROM ${pipelineConfig.source_table}`;
  const ctx = resolveDatabricksContext(pipelineConfig);
  const warehouseId = ctx.httpPath.split("/").pop();

  logInfo(`Querying Databricks`, { table: pipelineConfig.source_table, profile: ctx.auth });
  const token =
    ctx.auth === "pat" ? ctx.pat : await getDatabricksOAuthToken();
  const res = await fetch(`https://${ctx.host}/api/2.0/sql/statements`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      warehouse_id: warehouseId,
      statement,
      wait_timeout: "50s",
      disposition: "INLINE",
    }),
  });

  if (!res.ok) {
    const txt = await res.text();
    throw new Error(`Databricks query failed (${res.status}): ${txt.slice(0, 300)}`);
  }

  const payload = await res.json();
  if (payload.status?.state !== "SUCCEEDED") {
    throw new Error(`Databricks statement status: ${payload.status?.state || "UNKNOWN"}`);
  }

  const rows = payload.result?.data_array || [];
  logInfo(`Fetched ${rows.length} row(s) from Databricks.`);
  return rows;
}

async function getExistingKeys(sb, targetTable, conflictKey) {
  const values = new Set();
  let offset = 0;
  const batch = 1000;

  while (true) {
    const res = await sb.get(
      targetTable,
      { select: conflictKey },
      { Range: `${offset}-${offset + batch - 1}` }
    );

    if (![200, 206].includes(res.status)) {
      throw new Error(`Failed to read existing keys: ${res.status}`);
    }

    const data = await res.json();
    if (!data.length) break;
    for (const row of data) {
      const v = row[conflictKey];
      if (v != null) values.add(String(v));
    }
    if (data.length < batch) break;
    offset += batch;
  }
  return values;
}

async function saveEtlLog(logData) {
  if (!ETL_LOGS_BUCKET) {
    console.warn("ETL_LOGS_BUCKET not set, skipping S3 log.");
    return null;
  }

  const ts = logData.sync_timestamp || new Date().toISOString();
  const datePrefix = ts.slice(0, 10);
  const pipelineName = logData.pipeline_name || "unknown";
  const key = `${pipelineName}/${datePrefix}/${ts.replace(/:/g, "-")}.json`;

  try {
    await s3Client.send(
      new PutObjectCommand({
        Bucket: ETL_LOGS_BUCKET,
        Key: key,
        Body: JSON.stringify(logData, null, 2),
        ContentType: "application/json",
      })
    );
    const uri = `s3://${ETL_LOGS_BUCKET}/${key}`;
    logInfo("ETL log saved", { uri });
    return uri;
  } catch (err) {
    console.error("Failed to save ETL log to S3", err);
    return null;
  }
}

async function createRunRecord(sb, pipelineName, status, payload = {}) {
  const row = {
    pipeline_name: pipelineName,
    status,
    source_table: payload.source_table || null,
    target_table: payload.dest_table || null,
    inserted_count: payload.inserted || 0,
    updated_count: payload.updated || 0,
    skipped_count: payload.skipped || 0,
    error_message: payload.error_message || null,
    started_at: payload.started_at || new Date().toISOString(),
    finished_at: payload.finished_at || new Date().toISOString(),
    duration_ms: payload.duration_ms || null,
    s3_log_uri: payload.s3_log_uri || null
  };
  try {
    const res = await sb.post("etl_runs", [row], null);
    if (![200, 201].includes(res.status)) {
      const txt = await res.text();
      console.warn(`etl_runs insert skipped (${res.status}): ${txt.slice(0, 200)}`);
    }
  } catch (err) {
    console.warn(`etl_runs insert skipped: ${String(err.message || err)}`);
  }
}

async function runSyncForPipeline(event) {
  const startedAt = new Date();
  const pipelineConfig = getPipelineConfig(event);
  const pipelineName = pipelineConfig.pipeline_name;
  const syncStart = new Date();
  const now = syncStart.toISOString();
  const sourceTable = pipelineConfig.source_table;
  const targetTable = pipelineConfig.target_table;
  const conflictKey = pipelineConfig.conflict_key;
  const batchSize = pipelineConfig.batch_size || BATCH_SIZE;
  const writeMode = pipelineConfig.write_mode || "insert_only";

  const { baseUrl, serviceKey } = resolveSupabaseForPipeline(pipelineConfig);
  const sb = makeSupabaseClient(baseUrl, serviceKey);

  const tableRes = await sb.get(targetTable, {
    select: conflictKey,
    limit: "1",
  });
  if (tableRes.status !== 200) {
    throw new Error(`Cannot access '${targetTable}': ${tableRes.status}`);
  }

  const wmRes = await sb.get(WATERMARK_TABLE, {
    select: "table_name",
    limit: "1",
  });
  if (wmRes.status !== 200) {
    throw new Error(
      `Cannot access '${WATERMARK_TABLE}': ${wmRes.status}. Create ETL tables in this Supabase project.`
    );
  }

  const existing = await getExistingKeys(sb, targetTable, conflictKey);
  const countBefore = existing.size;
  logInfo(`Found ${countBefore} existing records in Supabase.`, { targetTable, pipelineName });

  let dbxRows = await fetchDatabricks(pipelineConfig);
  if (!dbxRows.length) {
    return { status: "no_data", inserted: 0, total: countBefore, pipeline_name: pipelineName };
  }
  dbxRows = deduplicateRows(dbxRows, pipelineConfig);

  const { newRows, allRows } = splitRowsByMode(dbxRows, pipelineConfig, existing, now);
  logInfo("Sync split", {
    pipeline: pipelineName,
    new: newRows.length,
    skipped: dbxRows.length - newRows.length
  });

  let inserted = 0;
  let updated = 0;
  const rowsToWrite = writeMode === "upsert" ? allRows : newRows;
  for (let i = 0; i < rowsToWrite.length; i += batchSize) {
    const batch = rowsToWrite.slice(i, i + batchSize);
    const res =
      writeMode === "upsert"
        ? await sb.upsert(targetTable, batch, conflictKey)
        : await sb.post(targetTable, batch, null);
    if (![200, 201].includes(res.status)) {
      const txt = await res.text();
      throw new Error(`Write failed: ${res.status} ${txt.slice(0, 300)}`);
    }
    if (writeMode === "upsert") {
      const insertedInBatch = batch.filter(
        (r) => r[conflictKey] != null && !existing.has(String(r[conflictKey]))
      ).length;
      inserted += insertedInBatch;
      updated += batch.length - insertedInBatch;
      for (const r of batch) {
        if (r[conflictKey] != null) {
          existing.add(String(r[conflictKey]));
        }
      }
    } else {
      inserted += batch.length;
    }
    logInfo(`Written batch ${Math.floor(i / batchSize) + 1}`, { mode: writeMode, size: batch.length });
  }

  const postKeys = await getExistingKeys(sb, targetTable, conflictKey);
  const countAfter = postKeys.size;
  const syncTime = new Date().toISOString();

  const upsertRes = await sb.upsert(
    WATERMARK_TABLE,
    [{ table_name: pipelineName, last_timestamp: syncTime }],
    "table_name"
  );
  if (![200, 201].includes(upsertRes.status)) {
    console.warn(`Watermark update failed: ${upsertRes.status}`);
  }

  const detail = {
    pipeline_name: pipelineName,
    status: "success",
    inserted,
    updated,
    total: countAfter,
    source_table: sourceTable,
    dest_table: targetTable,
    write_mode: writeMode,
    conflict_key: conflictKey,
    source_count: dbxRows.length,
    dest_count_before: countBefore,
    dest_count_after: countAfter,
    skipped: dbxRows.length - rowsToWrite.length,
    sync_timestamp: syncTime,
    inserted_records: newRows
      .slice(0, 200)
      .map((r) => ({ [conflictKey]: r[conflictKey] })),
  };
  const s3LogUri = await saveEtlLog(detail);
  const finishedAt = new Date();
  await createRunRecord(sb, pipelineName, "success", {
    ...detail,
    started_at: startedAt.toISOString(),
    finished_at: finishedAt.toISOString(),
    duration_ms: finishedAt.getTime() - startedAt.getTime(),
    s3_log_uri: s3LogUri
  });

  return {
    pipeline_name: pipelineName,
    status: "success",
    inserted,
    updated,
    total: countAfter,
    sync_timestamp: syncTime,
  };
}

exports.lambdaHandler = async (event) => {
  logInfo("Event received", event);
  try {
    const result = await runSyncForPipeline(event || {});
    logInfo("Sync complete", result);
    return {
      statusCode: 200,
      body: JSON.stringify(result),
    };
  } catch (error) {
    console.error("Sync failed", error);
    try {
      const pipelineName = event?.pipeline_name || process.env.DEFAULT_PIPELINE || "transportistas";
      const cfg = loadPipelinesConfig()[pipelineName];
      if (cfg) {
        const { baseUrl, serviceKey } = resolveSupabaseForPipeline(cfg);
        const sb = makeSupabaseClient(baseUrl, serviceKey);
        await createRunRecord(sb, pipelineName, "error", {
          error_message: String(error.message || error)
        });
      }
    } catch (inner) {
      console.error("Failed writing etl_runs error row", inner);
    }
    return {
      statusCode: 500,
      body: JSON.stringify({ status: "error", message: String(error.message || error) }),
    };
  }
};
